engine/access/rpc/backend/backend_network.go

```diff
@@ -80,6 +83,71 @@ func (b *backendNetwork) GetLatestProtocolStateSnapshot(_ context.Context) ([]by
 	return data, nil
 }
 
+// GetProtocolStateSnapshotByBlockID returns serializable Snapshot by blockID
+func (b *backendNetwork) GetProtocolStateSnapshotByBlockID(_ context.Context, blockID flow.Identifier) ([]byte, error) {
+	snapshotByBlockId := b.state.AtBlockID(blockID)
+	snapshotHeadByBlockId, err := snapshotByBlockId.Head()
+	if err != nil {
+		if errors.Is(err, state.ErrUnknownSnapshotReference) {
+			return nil, status.Errorf(codes.NotFound, "failed to get a valid snapshot: %v", err)
+		}
+
+		return nil, status.Errorf(codes.Internal, "failed to get a valid snapshot: %v", err)
+	}
+
+	snapshotByHeight := b.state.AtHeight(snapshotHeadByBlockId.Height)
+	snapshotHeadByHeight, err := snapshotByHeight.Head()
+	if err != nil {
+		if errors.Is(err, state.ErrUnknownSnapshotReference) {
+			return nil, status.Errorf(codes.NotFound, "failed to get a valid snapshot: %v", err)
```

Comment:
this should be an `InvalidArgument` error so they know the issue was with the request.
```suggestion
			return nil, status.Errorf(codes.InvalidArgument, "failed to retrieve snapshot for block by height: block not finalized")
```